integration/test/Test/Connection.hs

```diff
@@ -0,0 +1,403 @@
+  [alice, bob] <- forM [OwnDomain, OtherDomain] $ flip randomUser def
+  -- set up an initial "ignored" state
+  void $ postConnection bob alice >>= getBody 201
+  void $ putConnection alice bob "ignored" >>= getBody 200
+  assertConnectionStatus alice bob "ignored"
+
+  -- if Bob rescinds, Alice stays in "ignored"
+  void $ putConnection bob alice "cancelled" >>= getBody 200
+  assertConnectionStatus alice bob "ignored"
+
+  -- if Alice accepts, and Bob does not want to connect anymore, Alice
+  -- transitions to "sent"
+  void $ putConnection alice bob "accepted" >>= getBody 200
+  assertConnectionStatus alice bob "sent"
+
+testConnectFromBlocked :: HasCallStack => App ()
+testConnectFromBlocked = do
+  (alice, bob, one2oneId) <- createOne2OneConversation OwnDomain
+  bobId <- bob %. "qualified_id"
+
+  -- set up an initial "blocked" state
+  void $ postConnection bob alice >>= getBody 200
+  void $ putConnection alice bob "blocked" >>= getBody 200
+  assertConnectionStatus alice bob "blocked"
+  getConversation alice one2oneId `bindResponse` \resp ->
+    resp.status `shouldMatchInt` 403
+
+  -- If Bob sends a new connection request, Alice ignores it
+  void $ postConnection bob alice >>= getBody 200
+  assertConnectionStatus alice bob "blocked"
+
+  -- if Alice accepts (or sends a connection request), and Bob still
+  -- wants to connect, Alice transitions to "accepted"
+  void $ postConnection alice bob >>= getBody 200
+  assertConnectionStatus alice bob "accepted"
+  getConversation alice one2oneId `bindResponse` \resp -> do
+    resp.status `shouldMatchInt` 200
+    others <- resp.json %. "members.others" & asList
+    qIds <- for others (%. "qualified_id")
+    qIds `shouldMatchSet` [bobId]
+
+testSentFromBlocked :: HasCallStack => App ()
+testSentFromBlocked = do
+  [alice, bob] <- forM [OwnDomain, OtherDomain] $ flip randomUser def
+  -- set up an initial "blocked" state
+  void $ postConnection bob alice >>= getBody 201
+  void $ putConnection alice bob "blocked" >>= getBody 200
+  assertConnectionStatus alice bob "blocked"
+
+  -- if Bob rescinds, Alice stays in "blocked"
+  void $ putConnection bob alice "cancelled" >>= getBody 200
+  assertConnectionStatus alice bob "blocked"
+
+  -- if Alice accepts, and Bob does not want to connect anymore, Alice
+  -- transitions to "sent"
+  void $ putConnection alice bob "accepted" >>= getBody 200
+  assertConnectionStatus alice bob "sent"
+
+testCancel :: HasCallStack => App ()
+testCancel = do
+  [alice, bob] <- forM [OwnDomain, OtherDomain] $ flip randomUser def
+
+  void $ postConnection alice bob >>= getBody 201
+  assertConnectionStatus alice bob "sent"
+
+  void $ putConnection alice bob "cancelled" >>= getBody 200
+  assertConnectionStatus alice bob "cancelled"
+
+testConnectionLimits :: HasCallStack => App ()
+testConnectionLimits = do
+  let connectionLimit = 16
+
+  alice <- randomUser OwnDomain def
+  [charlie1, charlie2, charlie3, charlie4] <- replicateM 4 do
```

Comment:
We've got a helper to create multiple users: `createUsers $ replicate 4 OtherDomain`

Reply:
Iirc this does more than I'm allowed to do here. (To make the test replicate the old behaviour)